streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -16,10 +16,294 @@
+        final List<String> sourceTopics = new ArrayList<>();
+        applyToEachBuilder(b -> sourceTopics.addAll(b.sourceTopicCollection()));
+        return sourceTopics;
+    }
+
+    Pattern sourceTopicPattern() {
+        final StringBuilder patternBuilder = new StringBuilder();
+
+        applyToEachBuilder(b -> {
+            final String patternString = b.sourceTopicsPatternString();
+            if (patternString.length() > 0) {
+                patternBuilder.append(patternString).append("|");
+            }
+        });
+
+        if (patternBuilder.length() > 0) {
+            patternBuilder.setLength(patternBuilder.length() - 1);
+            return Pattern.compile(patternBuilder.toString());
+        } else {
+            return EMPTY_ZERO_LENGTH_PATTERN;
+        }
+    }
+
+    public boolean usesPatternSubscription() {
+        return evaluateConditionIsTrueForAnyBuilders(InternalTopologyBuilder::usesPatternSubscription);
+    }
+
+    // Can be empty if app is started up with no Named Topologies, in order to add them on later
+    public boolean isEmpty() {
+        return builders.isEmpty();
+    }
+
+    public String topologyDescription() {
+        if (isEmpty()) {
+            return "";
+        }
+        final StringBuilder sb = new StringBuilder();
+
+        applyToEachBuilder(b -> {
+            sb.append(b.describe().toString());
+        });
+
+        return sb.toString();
+    }
+
+    public final void buildAndRewriteTopology() {
+        applyToEachBuilder(builder -> {
+            builder.rewriteTopology(config);
+            builder.buildTopology();
+
+            // As we go, check each topology for overlap in the set of input topics/patterns
+            final int numInputTopics = allInputTopics.size();
+            final List<String> inputTopics = builder.fullSourceTopicNames();
+            final Collection<String> inputPatterns = builder.allSourcePatternStrings();
+
+            final int numNewInputTopics = inputTopics.size() + inputPatterns.size();
+            allInputTopics.addAll(inputTopics);
+            allInputTopics.addAll(inputPatterns);
+            if (allInputTopics.size() != numInputTopics + numNewInputTopics) {
+                inputTopics.retainAll(allInputTopics);
+                inputPatterns.retainAll(allInputTopics);
+                inputTopics.addAll(inputPatterns);
+                log.error("Tried to add the NamedTopology {} but it had overlap with other input topics: {}", builder.namedTopology(), inputTopics);
+                throw new TopologyException("Named Topologies may not subscribe to the same input topics or patterns");
+            }
+
+            final ProcessorTopology globalTopology = builder.buildGlobalStateTopology();
+            if (globalTopology != null) {
+                if (builder.namedTopology() != null) {
+                    throw new IllegalStateException("Global state stores are not supported with Named Topologies");
```

Comment:
Okay now I see why we do not count named topologies for global stores for caching. Is it final, or just not supported within pt.2?

Reply:
It's not likely to be compatible in the first phase, but I think we would want it to be fully supported in the end, ie as part of the KIP once we have everything worked out and feeling ready for that